### PR TITLE
Update to MP renamings

### PR DESCRIPTION
**Pull request by blegat**

### CODE REVIEW

**Comment by blegat** [comment]
@kalmarek `PermutationGroups.SymmetricGroup` is gone, what is the replacement ?

**Comment by kalmarek** [comment]
this thing came from AA previously;

At the moment probably it's best to define
```julia
julia> using PermutationGroups
[ Info: Precompiling PermutationGroups [8bc5a954-2dfc-11e9-10e6-cd969bffa420]

julia> symmetric_group(n) = PermGroup(Perm([2,1]), Perm([2:n;1]))
symmetric_group (generic function with 1 method)

julia> PermutationGroups.order(symmetric_group(4))
24

julia> collect(symmetric_group(3))
6-element Vector{Permutation{Perm{UInt16}, PermGroup{Perm{UInt16}, …}}}:
 ()
 (2,3)
 (1,2)
 (1,2,3)
 (1,3,2)
 (1,3)

```

Do you think `PermutationGroups.jl` should provide `symmetric_group` and `alternating_group`?

**Comment by blegat** [comment]
@kalmarek Now I have
```
ERROR: MethodError: no method matching PermutationGroups.Perms.Perm(::Int64)

Closest candidates are:
  PermutationGroups.Perms.Perm(::AbstractVector{<:Integer})
   @ PermutationGroups ~/.julia/packages/PermutationGroups/Ox6EV/src/Perms/perm_images.jl:49
  PermutationGroups.Perms.Perm(::AbstractVector{<:Integer}, ::Any)
   @ PermutationGroups ~/.julia/packages/PermutationGroups/Ox6EV/src/Perms/perm_images.jl:49

Stacktrace:
  [1] one(#unused#::CondensedMatterSOS.KleinPermGroup)
    @ CondensedMatterSOS ~/.julia/dev/CondensedMatterSOS/src/symmetry.jl:96
  [2] one(L::Lattice1Group)
    @ CondensedMatterSOS ~/.julia/dev/CondensedMatterSOS/src/symmetry.jl:189
  [3] conjugacy_classes_orbit(G::Lattice1Group)
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/cmmatrix.jl:44
  [4] conjugacy_classes(G::Lattice1Group)
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/cmmatrix.jl:41
  [5] SymbolicWedderburn.Characters.CharacterTable(R::Type{Rational{Int64}}, G::Lattice1Group)
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:82
  [6] symmetry_adapted_basis(T::Type, G::Lattice1Group, action::Action, basis::MonomialBasis{CondensedMatterSOS.SpinMonomial, Vector{CondensedMatterSOS.SpinMonomial}}, S::Type; semisimple::Bool)
    @ SymbolicWedderburn ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/sa_basis.jl:163
```

**Comment by blegat** [comment]
@kalmarek I now get
```
ERROR: DomainError with 6:
6 is not a square modulo 13
Stacktrace:
  [1] sqrtmod(n::Int64, q::Int64)
    @ SymbolicWedderburn.Characters.FiniteFields ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:109
  [2] sqrt
    @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:103 [inlined]
  [3] normalize!(chtbl::SymbolicWedderburn.Characters.CharacterTable{Lattice1Group, SymbolicWedderburn.Characters.FiniteFields.GF{13}, PermutationGroups.Orbit{CondensedMatterSOS.DirectSum}})
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:141
  [4] SymbolicWedderburn.Characters.CharacterTable(Fp::Type{SymbolicWedderburn.Characters.FiniteFields.GF{13}}, G::Lattice1Group, cclasses::Vector{PermutationGroups.Orbit{CondensedMatterSOS.DirectSum}})
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:73
  [5] SymbolicWedderburn.Characters.CharacterTable(R::Type{Rational{Int64}}, G::Lattice1Group, cclasses::Vector{PermutationGroups.Orbit{CondensedMatterSOS.DirectSum}})
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:83
  [6] SymbolicWedderburn.Characters.CharacterTable(R::Type{Rational{Int64}}, G::Lattice1Group)
    @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:82
  [7] symmetry_adapted_basis(T::Type, G::Lattice1Group, action::Action, basis::MonomialBasis{CondensedMatterSOS.SpinMonomial, Vector{CondensedMatterSOS.SpinMonomial}}, S::Type; semisimple::Bool)
    @ SymbolicWedderburn ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/sa_basis.jl:163
```
Is this a bug in CondensedMatterSOS ? It could be that my group is not satisfying the axioms of a group but we are checking that now

**Comment by kalmarek** [comment]
@blegat where is this group defined?

**Comment by blegat** [comment]
Here:
https://github.com/blegat/CondensedMatterSOS.jl/blob/bl/rename/src/symmetry.jl

**Comment by blegat** [comment]
Still the same error it seems.

**Comment by kalmarek** [comment]
hmm
```julia
julia> G = Lattice1Group(2)
Lattice1Group(CondensedMatterSOS.CyclicGroup(2), CondensedMatterSOS.KleinPermGroup())

julia> SymbolicWedderburn.Characters.irreducible_characters(G)
ERROR: DomainError with 6:
6 is not a square modulo 13
Stacktrace:
 [1] sqrtmod(n::Int64, q::Int64)
   @ SymbolicWedderburn.Characters.FiniteFields ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:109
 [2] sqrt
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:103 [inlined]
 [3] normalize!(chtbl::SymbolicWedderburn.Characters.CharacterTable{Lattice1Group, SymbolicWedderburn.Characters.FiniteFields.GF{13}, PermutationGroups.Orbit{CondensedMatterSOS.DirectSumElem{CondensedMatterSOS.CyclicElem, CondensedMatterSOS.KleinPermElement}}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:141
 [4] SymbolicWedderburn.Characters.CharacterTable(Fp::Type{SymbolicWedderburn.Characters.FiniteFields.GF{13}}, G::Lattice1Group, cclasses::Vector{PermutationGroups.Orbit{CondensedMatterSOS.DirectSumElem{CondensedMatterSOS.CyclicElem, CondensedMatterSOS.KleinPermElement}}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:73
 [5] SymbolicWedderburn.Characters.CharacterTable(R::Type{Rational{Int64}}, G::Lattice1Group, cclasses::Vector{PermutationGroups.Orbit{CondensedMatterSOS.DirectSumElem{CondensedMatterSOS.CyclicElem, CondensedMatterSOS.KleinPermElement}}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:83
 [6] irreducible_characters
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:41 [inlined]
 [7] irreducible_characters
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:33 [inlined]
 [8] irreducible_characters(G::Lattice1Group)
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:33
 [9] top-level scope
   @ REPL[21]:1

julia> SymbolicWedderburn.Characters.irreducible_characters(G.K)
ERROR: DomainError with 6:
6 is not a square modulo 13
Stacktrace:
 [1] sqrtmod(n::Int64, q::Int64)
   @ SymbolicWedderburn.Characters.FiniteFields ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:109
 [2] sqrt
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:103 [inlined]
 [3] normalize!(chtbl::SymbolicWedderburn.Characters.CharacterTable{CondensedMatterSOS.KleinPermGroup, SymbolicWedderburn.Characters.FiniteFields.GF{13}, PermutationGroups.Orbit{CondensedMatterSOS.KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:141
 [4] SymbolicWedderburn.Characters.CharacterTable(Fp::Type{SymbolicWedderburn.Characters.FiniteFields.GF{13}}, G::CondensedMatterSOS.KleinPermGroup, cclasses::Vector{PermutationGroups.Orbit{CondensedMatterSOS.KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:73
 [5] SymbolicWedderburn.Characters.CharacterTable(R::Type{Rational{Int64}}, G::CondensedMatterSOS.KleinPermGroup, cclasses::Vector{PermutationGroups.Orbit{CondensedMatterSOS.KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:83
 [6] irreducible_characters
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:41 [inlined]
 [7] irreducible_characters
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:33 [inlined]
 [8] irreducible_characters(G::CondensedMatterSOS.KleinPermGroup)
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:33
 [9] top-level scope
   @ REPL[22]:1

julia> order(G.K)
24
```
Since all groups of order up to 30 are tested here:
https://github.com/kalmarek/SymbolicWedderburn.jl/blob/master/test/dixon.jl#L229
I'd assume something is wrong with the implementation of `KleinPermGroup`. I didn't check all the formulas while rewriting.

**Comment by kalmarek** [comment]
TLDR: bug was in `gens(::KleinPermGroup)`, fix as suggested;

There are also two fixes for botched `deepcopy_internal` I recommend applying.

<details>

<summary>Longer version if you wish...</summary>

```julia
julia> K = KleinPermGroup()
KleinPermGroup()

julia> Characters.CharacterTable(Rational{Int}, K)
ERROR: DomainError with 6:
6 is not a square modulo 13
Stacktrace:
 [1] sqrtmod(n::Int64, q::Int64)
   @ SymbolicWedderburn.Characters.FiniteFields ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:109
 [2] sqrt
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:103 [inlined]
 [3] normalize!(chtbl::CharacterTable{KleinPermGroup, SymbolicWedderburn.Characters.FiniteFields.GF{13}, Orbit{KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:141
 [4] CharacterTable(Fp::Type{SymbolicWedderburn.Characters.FiniteFields.GF{13}}, G::KleinPermGroup, cclasses::Vector{Orbit{KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:73
 [5] CharacterTable(R::Type{Rational{Int64}}, G::KleinPermGroup, cclasses::Vector{Orbit{KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:83
 [6] CharacterTable(R::Type{Rational{Int64}}, G::KleinPermGroup)
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:82
 [7] top-level scope
   @ REPL[4]:1
```
the error that we've seen.

```julia
julia> p = Characters.dixon_prime(K)
13

julia> Characters.CharacterTable(Characters.FiniteFields.GF{p}, K)
ERROR: DomainError with 6:
6 is not a square modulo 13
Stacktrace:
 [1] sqrtmod(n::Int64, q::Int64)
   @ SymbolicWedderburn.Characters.FiniteFields ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:109
 [2] sqrt
   @ ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/gf.jl:103 [inlined]
 [3] normalize!(chtbl::CharacterTable{KleinPermGroup, SymbolicWedderburn.Characters.FiniteFields.GF{13}, Orbit{KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:141
 [4] CharacterTable(Fp::Type{SymbolicWedderburn.Characters.FiniteFields.GF{13}}, G::KleinPermGroup, cclasses::Vector{Orbit{KleinPermElement}})
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:73
 [5] CharacterTable(Fp::Type{SymbolicWedderburn.Characters.FiniteFields.GF{13}}, G::KleinPermGroup)
   @ SymbolicWedderburn.Characters ~/.julia/packages/SymbolicWedderburn/hBhTJ/src/Characters/character_tables.jl:61
 [6] top-level scope
   @ REPL[6]:1
```
actually happens when computing over the finite field. Let's investigate further.
There are 15 groups of order 24:

```julia
julia> SWpath = joinpath(dirname(dirname(pathof(SymbolicWedderburn))),)
"/home/kalmar/.julia/packages/SymbolicWedderburn/hBhTJ/test"

julia> include(joinpath(SWpath, "test", "smallgroups.jl"));

julia> length(SmallPermGroups[24])
15
```
one of them should be isomorphic to our `K`. Which one? Let's gather some candidates

```julia
julia> K_ords = sort!(vec(map(g->order(Int, g), K))))
24-element Vector{Int64}:
 1
 2
 2
 2
 2
 2
 2
 2
 ⋮
 3
 4
 4
 4
 4
 4
 4

julia> candidates = [i for (i, G) in pairs(SmallPermGroups[24]) if
           sort!(map(g->order(Int, g), G)) == K_ords]
1-element Vector{Int64}:
 12

julia> K_perm = SmallPermGroups[24][first(candidates)]
Permutation group on 4 generators of order 24 generated by
 (1,2)(3,13)(4,8)(5,7)(6,9)(10,21)(11,20)(12,16)(14,18)(15,17)(19,24)(22,23)
 (1,3,9)(2,6,13)(4,11,23)(5,19,17)(7,15,24)(8,22,20)(10,18,12)(14,21,16)
 (1,4)(2,7)(3,10)(5,12)(6,14)(8,16)(9,17)(11,19)(13,20)(15,22)(18,23)(21,24)
 (1,5)(2,8)(3,11)(4,12)(6,15)(7,16)(9,18)(10,19)(13,21)(14,22)(17,23)(20,24)

```
Clearly isomorphism should preserve orders, and there's only one group with the same set of orders as `K`. As expected though we can compute its characters just fine over the same `p`:

```
julia> let p = p
           Characters.CharacterTable(Characters.FiniteFields.GF{p}, K_perm)
       end
Character table of PermGroup( (1,2)(3,13)(4,8)(5,7)(6,9)(10,21)(11,20)(12,16)(14,18)(15,17)(19,24)(22,23), (1,3,9)(2,6,13)(4,11,23)(5,19,17)(7,15,24)(8,22,20)(10,18,12)(14,21,16), (1,4)(2,7)(3,10)(5,12)(6,14)(8,16)(9,17)(11,19)(13,20)(15,22)(18,23)(21,24), (1,5)(2,8)(3,11)(4,12)(6,15)(7,16)(9,18)(10,19)(13,21)(14,22)(17,23)(20,24) ) over SymbolicWedderburn.Characters.FiniteFields.GF{13}
       1    2    3    4    5
───┬────────────────────────
χ₁ │ 2₁₃  0₁₃ 12₁₃  2₁₃  0₁₃
χ₂ │ 3₁₃  1₁₃  0₁₃ 12₁₃ 12₁₃
χ₃ │ 1₁₃  1₁₃  1₁₃  1₁₃  1₁₃
χ₄ │ 1₁₃ 12₁₃  1₁₃  1₁₃ 12₁₃
χ₅ │ 3₁₃ 12₁₃  0₁₃ 12₁₃  1₁₃

```

Ok, since we get a strange error while normalizing "characters" my guess is that something is wrong with conjugacy classes of `K`. Are the orders equal inside each class?
```julia
julia> [order.(Int, c) for c in Characters.conjugacy_classes(K)]
7-element Vector{Vector{Int64}}:
 [1]
 [2, 2, 2]
 [3, 3]
 [2, 2, 2]
 [2, 2, 2]
 [4, 4, 4, 4, 4, 4]
 [3, 3, 3, 3, 3, 3]

julia> [order.(Int, c) for c in Characters.conjugacy_classes(K_perm)]
5-element Vector{Vector{Int64}}:
 [1]
 [2, 2, 2, 2, 2, 2]
 [3, 3, 3, 3, 3, 3, 3, 3]
 [2, 2, 2]
 [4, 4, 4, 4, 4, 4]
```
yes they are, but it seems that some elements which should be conjugated are not?

Lets take a permutation representation of `K` acting on itself via right multiplication:

```julia
julia> S = let elts = collect(K)
           idict = Dict(e=>i for (i,e) in pairs(vec(elts)))

           map(gens(K)) do g
               Perm([idict[h*g] for h in vec(elts)])
           end
       end;

julia> PermGroup(S)
Permutation group on 2 generators generated by
 (1,3)(2,4)(5,6)(7,15)(8,16)(9,13)(10,14)(11,18)(12,17)(19,21)(20,22)(23,24)
 (1,4,5)(2,3,6)(7,16,23)(8,15,24)(9,18,20)(10,17,19)(11,13,22)(12,14,21)

julia> order(Int, ans)
6
```
only two generators and order 6! we have a solution:

```julia
julia> gens(K)
2-element Vector{KleinPermElement}:
 KleinPermElement((1,2), KleinElement(0))
 KleinPermElement((1,2,3), KleinElement(0))
```
these surely don't generate `K` :D

----

With
```julia
function GroupsCore.gens(::KleinPermGroup)
    Kid = one(KleinGroup())
    elts = [KleinPermElement(g, Kid) for g in __symmetric_grp_gens(3)]
    Pid = one(first(__symmetric_grp_gens(3)))
    append!(elts, KleinPermElement(Pid, k) for k in gens(KleinGroup()))
    return elts
end
```
I get
```julia
julia> Characters.CharacterTable(Rational{Int}, K)
Character table of KleinPermGroup() over Cyclotomics.Cyclotomic{Rational{Int64}, SparseArrays.SparseVector{Rational{Int64}, Int64}}
        1     2     3     4     5
───┬─────────────────────────────
χ₁ │ 2//1  0//1 -1//1  2//1  0//1
χ₂ │ 3//1  1//1  0//1 -1//1 -1//1
χ₃ │ 1//1  1//1  1//1  1//1  1//1
χ₄ │ 1//1 -1//1  1//1  1//1 -1//1
χ₅ │ 3//1 -1//1  0//1 -1//1  1//1

julia> Characters.CharacterTable(Rational{Int}, K_perm)
Character table of PermGroup( (1,2)(3,13)(4,8)(5,7)(6,9)(10,21)(11,20)(12,16)(14,18)(15,17)(19,24)(22,23), (1,3,9)(2,6,13)(4,11,23)(5,19,17)(7,15,24)(8,22,20)(10,18,12)(14,21,16), (1,4)(2,7)(3,10)(5,12)(6,14)(8,16)(9,17)(11,19)(13,20)(15,22)(18,23)(21,24), (1,5)(2,8)(3,11)(4,12)(6,15)(7,16)(9,18)(10,19)(13,21)(14,22)(17,23)(20,24) ) over Cyclotomics.Cyclotomic{Rational{Int64}, SparseArrays.SparseVector{Rational{Int64}, Int64}}
        1     2     3     4     5
───┬─────────────────────────────
χ₁ │ 2//1  0//1 -1//1  2//1  0//1
χ₂ │ 3//1  1//1  0//1 -1//1 -1//1
χ₃ │ 1//1  1//1  1//1  1//1  1//1
χ₄ │ 1//1 -1//1  1//1  1//1 -1//1
χ₅ │ 3//1 -1//1  0//1 -1//1  1//1
```
</details>

File: src/symmetry.jl [hunk 1]
```diff
@@ -93,18 +108,16 @@ function PermutationGroups.order(el::KleinPermElement)
...
-Base.one(::Union{KleinPermGroup, KleinPermElement}) = KleinPermElement(Perm(3), one(KleinGroup()))
+Base.one(::Union{KleinPermGroup, KleinPermElement}) = KleinPermElement(one(G3), one(KleinGroup()))
 # See https://github.com/blegat/CondensedMatterSOS.jl/pull/31#issuecomment-1717665659
-symmetric_group(n) = PermutationGroups.PermGroup(PermutationGroups.Perm([2,1]), PermutationGroups.Perm([2:n;1]))
-function PermutationGroups.gens(::KleinPermGroup)
-    els = [KleinPermElement(Perm(3), KleinElement(1))]
-    for g in gens(symmetric_group(3))
+function PermutationGroups.gens(g::KleinPermGroup)
+    els = [one(g)]
+    for g in gens(G3)
```
Comment (kalmarek) [inline src/symmetry.jl]: @blegat this is where the bug was introduced;
The old version generated `KleinPermGroup`, since `Sym(3)` acts transively on non-trivial `KleinElement`s, the new one doesn't since `KleinElement(0)` is fixed under `Sym(3)`.

I just updated the function without thinking ;)

File: src/symmetry.jl [hunk 2]
```diff
@@ -21,40 +48,52 @@ function Base.:*(a::KleinElement, b::KleinElement)
...
 struct KleinPermElement <: GroupElement
-    p::Perm{Int}
+    p::PermutationGroups.Perm{UInt16}
     k::KleinElement
 end
+
+function GroupsCore.gens(::KleinPermGroup)
+    Kid = one(KleinGroup())
+    return [KleinPermElement(g, Kid) for g in __symmetric_grp_gens(3)]
+end
```
Comment (kalmarek) [inline src/symmetry.jl]:
```suggestion
function GroupsCore.gens(::KleinPermGroup)
    sym3_gens = __symmetric_grp_gens(3)
    Kid = one(KleinGroup())
    elts = [KleinPermElement(g, Kid) for g in sym3_gens]
    Pid = one(first(sym3_gens))
    append!(elts, KleinPermElement(Pid, k) for k in gens(KleinGroup()))
    return elts
end
```

File: src/symmetry.jl [hunk 3]
```diff
@@ -64,154 +103,157 @@ function Base.:(==)(a::KleinPermElement, b::KleinPermElement)
...
+function Base.deepcopy_internal(g::KleinPermElement, dict::IdDict)
+    p = if haskey(dict, objectid(g.p))
+        dict[objectid(g.p)]
     else
-        el, st = el_st
-        return KleinPermElement(P3[el[1]], KleinElement(el[2])), st
+        Base.deepcopy_internal(g.p, dict)
     end
+    return KleinPermElement(p, g.k)
 end
```
Comment (kalmarek) [inline src/symmetry.jl]: I think I've botched both `deepcopy_internal`s. Here's the corrected version:
```julia
function Base.deepcopy_internal(g::KleinPermElement, dict::IdDict)
    haskey(dict, g) && return dict[g]

    h = KleinPermElement(Base.deepcopy_internal(g.p, dict), g.k)
    dict[g] = h
    return h
end
```

File: src/symmetry.jl [hunk 4]
```diff
@@ -64,154 +103,157 @@ function Base.:(==)(a::KleinPermElement, b::KleinPermElement)
...
-        push!(els, DirectSum(one(CyclicGroup(L.n)), g))
+
+function Base.deepcopy_internal(g::DirectSumElem, dict::IdDict)
+    h = if isbits(g.h)
+        g.h
+    elseif haskey(dict, objectid(g.h))
+        dict[objectid(g.h)]
+    else
+        Base.deepcopy_internal(g.h, dict)
     end
```
Comment (kalmarek) [inline src/symmetry.jl]: and again:
```julia
function Base.deepcopy_internal(g::DirectSumElem, dict::IdDict)
    haskey(dict, g) && return g
    h = if isbits(g.h)
        g.h
    elseif haskey(dict, g.h)
        dict[g.h]
    else
        Base.deepcopy_internal(g.h, dict)
    end
    k = if isbits(g.k)
        g.k
    elseif haskey(dict, g.k)
        dict[g.k]
    else
        Base.deepcopy_internal(g.k, dict)
    end
    new_g = DirectSumElem(h, k)
    dict[g] = new_g
    return new_g
end
```
